differential/diffAtlas_utils.R

```diff
@@ -0,0 +1,327 @@
+  assaydata <- data.frame(do.call(rbind, lapply(metaRows, function(x){ x[! names(x) %in% all_colnames] <- NA; x[all_colnames]})))
+  
+  # Subset assay data to only the groups we'll be making contrasts from
+  
+  all_contrast_assaygroups <- unique(unlist(lapply(contrasts, function(x) c(x@reference_assay_group_id, x@test_assay_group_id))))
+  assaydata <- assaydata[assaydata$assay_group_id %in% all_contrast_assaygroups,]
+  
+  # Get covariates from contrasts (would it really be so hard to store it next to the assays?)
+  
+  if (any(unlist(lapply(contrasts, function(x) length(batch_effects(x)) > 0)))){
+    meta_from_contrasts <- unique(do.call(rbind, lapply(contrasts, function(cont){
+      if (length(batch_effects(cont)) > 0){
+        if (twocolor){
+          stop( "Cannot handle batch effects for 2-colour microarray data." )
+        }
+        do.call(rbind, lapply(batch_effects(cont), function(x){
+          do.call(rbind, lapply(names(x@batches), function(y) data.frame(id = x@batches[[y]], variable = x@effect_name, value = y  )))
+        }   ))
+      }
+    })))
+    if (!(is.null(meta_from_contrasts))){
+      assaydata <- merge(assaydata, dcast(meta_from_contrasts, id ~ variable), by.x = 'AssayName', by.y = 'id')
+    }
+  }
+  
+  rownames(assaydata) <- assaydata$AssayName
+
+  # For two-color add separate cleaned AssayName and BioRepName columns
+  
+  if (twocolor){
+    assaydata$AssayNameNoCy <- sub( ".Cy\\d$", "", assaydata$AssayName)
+    assaydata$BioRepNameNoCy <- sub( ".Cy\\d$", "", assaydata$BioRepName)
+    assaydata$dye <- sub('.*(Cy\\d+)', '\\1', assaydata$AssayName)
+  }
+  
+  # Make sure colnames are R-safe
+
+  colnames(assaydata) <- make.names(colnames(assaydata))
+
+  assaydata
+}
+
+# Make a simple contrasts table
+
+make_exp_contrast_table <- function(analytics){
+  
+  conts <- atlas_contrasts( analytics )
+  
+  conts <- data.frame(do.call(
+    rbind, 
+    lapply(conts, function(x){ 
+      unlist(
+        structure(
+          lapply(slotNames(x), function(y){ 
+            slot_val <- slot(x,y)
+            if (y == 'batch_effects' && length(slot_val) > 0){
+              paste(make.names(unlist(lapply(slot_val, function(z){ z@effect_name }))), collapse=' + ')
+            }else{
+              slot_val
+            }
+          }), 
+          names = slotNames(x))
+      )
+    })
+  ), stringsAsFactors = FALSE)
+  
+  # Make the formula we need for each contrast based on any batch effects present
+  
+  conts$formula <- apply(conts, 1, function(x){
+    formula <- '~ 0 + assay_group_id'
+    if ('batch_effects' %in% names(x) && !is.na(x['batch_effects'])){
+      formula <- paste(formula, x['batch_effects'], sep = ' + ')
+    }
+    formula
+  })
+  
+  conts
+}
+
+# Read a normalised expression table, log fold change table or mean intensities table
+
+read_exp_data_table <- function(expAcc, atlasProcessingDirectory, analytics, experiment, type = NULL, dataFilename = NULL){
```

Comment:
I might be wrong, but it looks to me that the parameter `type` must be always defined
```suggestion
read_exp_data_table <- function(expAcc, atlasProcessingDirectory, analytics, experiment, type, dataFilename = NULL){
```

Reply:
good shout